### PR TITLE
Manually add contacts to target groups

### PR DESCRIPTION
**Pull request by RainbowBunchie**
-   [x] Add changeset (if necessary)

Adds the possibility to manually add contacts to a target group. 

Screenshot 2024-06-19 at 09 37 33
Screenshot 2024-06-19 at 09 38 05

### CODE REVIEW

File: packages/admin/src/targetGroups/addContacts/AddContactsGridSelect.gql.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+`;
+
+export const assignedBrevoContactsQuery = gql`
+    query BrevoContactsGrid(
+        $offset: Int
+        $limit: Int
+        $email: String
+        $scope: EmailCampaignContentScopeInput!
+        $targetGroupId: ID
+        $onlyShowAssignedContactsOfTargetGroup: Boolean
```
Comment (thomasdax98) [inline packages/admin/src/targetGroups/addContacts/AddContactsGridSelect.gql.ts:21]: I'm unsure if I like this `onlyShowAssignedContactsOfTargetGroup` filter. It could also be two queries since the filter can't be set dynamically. 

But it's only a "slight objection"

**Review by thomasdax98 — APPROVED** [review]

File: packages/admin/src/targetGroups/addContacts/AddContactsGridSelect.tsx [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+    GQLAddBrevoContactsToTargetGroupMutationVariables,
+    GQLBrevoContactsGridQuery,
+    GQLBrevoContactsGridQueryVariables,
+    GQLRemoveBrevoContactFromTargetGroupMutation,
+    GQLRemoveBrevoContactFromTargetGroupMutationVariables,
+    GQLTargetGroupBrevoContactsListFragment,
+    namedOperations,
+} from "./AddContactsGridSelect.gql.generated";
+
+interface OpenSelectDialogToolbarProps {
```
Comment (johnnyomair) [inline packages/admin/src/targetGroups/addContacts/AddContactsGridSelect.tsx:41]: This should be called `AssignedContactsGridToolbar`

File: packages/admin/src/targetGroups/addContacts/AddContactsGridSelect.tsx [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+    openDialog: () => void;
```
Comment (johnnyomair) [inline packages/admin/src/targetGroups/addContacts/AddContactsGridSelect.tsx:42]: Optional: Callbacks usually start with `on`: `onOpenDialog`

File: .changeset/dirty-balloons-prove.md [hunk 4]
```diff
@@ -0,0 +1,6 @@
+---
+"@comet/brevo-admin": minor
+"@comet/brevo-api": minor
+---
+
+Contacts can now be manually assigned on the target group edit admin page additionally to the previously existing filter assignment
```
Comment (johnnyomair) [inline .changeset/dirty-balloons-prove.md:6]:
```suggestion
Allow manually assigning contacts to a target group

This is in addition to the existing automatic assignment via filters.
```

File: demo/api/src/db/migrations/Migration20240619092554.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import { Migration } from "@mikro-orm/migrations";
+
+export class Migration20240619092554 extends Migration {
```
[the remark this answers is not in the document]
Comment (RainbowBunchie) [reply demo/api/src/db/migrations/Migration20240619092554.ts:3]: Right now all migrations are in the demo - there is already a separate ticket to move the migrations into the module. I haven't figured out a way to combine the migrations with the comet migrations yet and to generate the columns for the scope and attributes dynamically. But this will be handled in upcoming ticket/PR - as soon as i have time for it

File: packages/admin/src/targetGroups/addContacts/AddContactsGridSelect.tsx (hunk 3 above)
Comment (RainbowBunchie) [reply packages/admin/src/targetGroups/addContacts/AddContactsGridSelect.tsx]: [b827bbf](https://github.com/vivid-planet/comet-brevo-module/pull/52/commits/b827bbf1c3d7f3b84162fcabea48f984e7c2ef4d)

File: packages/admin/src/targetGroups/addContacts/AddContactsGridSelect.tsx [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+    intl: IntlShape;
```
[the remark this answers is not in the document]
Comment (RainbowBunchie) [reply packages/admin/src/targetGroups/addContacts/AddContactsGridSelect.tsx]: [43cc7e8](https://github.com/vivid-planet/comet-brevo-module/pull/52/commits/43cc7e8e03dd0386185128a5131ee0d7653ec7a1)

File: .changeset/dirty-balloons-prove.md (hunk 4 above)
Comment (RainbowBunchie) [reply .changeset/dirty-balloons-prove.md]: [a7a7b56](https://github.com/vivid-planet/comet-brevo-module/pull/52/commits/a7a7b56c3bfdf92dcd41474c1cfe73a14d2b535c)

File: packages/admin/src/targetGroups/addContacts/AddContactsGridSelect.tsx (hunk 2 above)
Comment (RainbowBunchie) [reply packages/admin/src/targetGroups/addContacts/AddContactsGridSelect.tsx]: I removed the interface, since I removed the intl property the interface did not make sense anymore to be used in just 1 function for 1 property. 
[794246a](https://github.com/vivid-planet/comet-brevo-module/pull/52/commits/794246a9b7e475bd91404a2b60cff933004a142e)

File: packages/admin/src/targetGroups/addContacts/AddContactsGridSelect.gql.ts (hunk 1 above)
Comment (RainbowBunchie) [reply packages/admin/src/targetGroups/addContacts/AddContactsGridSelect.gql.ts]: [818cb40](https://github.com/vivid-planet/comet-brevo-module/pull/52/commits/818cb40799ed90aa33cfed1b9fd1c21f3dc1042e)

File: packages/admin/src/targetGroups/addContacts/AddContactsGridSelect.tsx [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+    } = useQuery<GQLBrevoContactsGridQuery, GQLBrevoContactsGridQueryVariables>(assignedBrevoContactsQuery, {
+        variables: {
+            offset: dataGridAssignedContactsProps.page * dataGridAssignedContactsProps.pageSize,
+            limit: dataGridAssignedContactsProps.pageSize,
+            email: dataGridAssignedContactsProps.filterModel?.quickFilterValues
+                ? dataGridAssignedContactsProps.filterModel?.quickFilterValues[0]
+                : undefined,
+            scope,
+            targetGroupId: id,
+            onlyShowAssignedContactsOfTargetGroup: true,
```
[the remark this answers is not in the document]
Comment (RainbowBunchie) [reply packages/admin/src/targetGroups/addContacts/AddContactsGridSelect.tsx]: I added a seperate query for using with targetGroupId
I changed the arg name to `onlyShowManuallyAssignedContacts` -> this is still necessary when wanting to differentiate between ALL contacts in a target group and just the manually assigned (which are shown in the datagrid in the screenshot). I think the naming for the argument was not good before.  

[818cb40](https://github.com/vivid-planet/comet-brevo-module/pull/52/commits/818cb40799ed90aa33cfed1b9fd1c21f3dc1042e)

File: demo/api/src/db/migrations/Migration20240619092554.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (RainbowBunchie) [reply demo/api/src/db/migrations/Migration20240619092554.ts:3]: As of today: yes, the migrations need to be done directly in the project unfortunately.

File: packages/admin/src/targetGroups/addContacts/AddContactsGridSelect.gql.ts (hunk 1 above)
Comment (johnnyomair) [reply packages/admin/src/targetGroups/addContacts/AddContactsGridSelect.gql.ts]: I don't think that's what @thomasdax98 was suggesting. Having a query arg `targetGroupId` is perfectly fine. I was thinking about a `manuallyAssignedBrevoContacts` query

File: .changeset/dirty-balloons-prove.md [hunk 8]
```diff
@@ -0,0 +1,7 @@
+---
+"@comet/brevo-admin": minor
+"@comet/brevo-api": minor
+---
+
+Allow manually assigning contacts to a target group
+This is in addition to the existing automatic assignment via filters.
```
Comment (johnnyomair) [inline .changeset/dirty-balloons-prove.md:7]:
```suggestion
Allow manually assigning contacts to a target group

This is in addition to the existing automatic assignment via filters.
```

The new line here is necessary to differentiate between title and description

**Review by johnnyomair — COMMENTED** [review]
Did you forget to push some changes? There are still some open concerns

File: packages/admin/src/targetGroups/addContacts/AddContactsGridSelect.gql.ts (hunk 1 above)
Comment (thomasdax98) [reply packages/admin/src/targetGroups/addContacts/AddContactsGridSelect.gql.ts]: > I don't think that's what @thomasdax98 was suggesting. Having a query arg `targetGroupId` is perfectly fine. I was thinking about a `manuallyAssignedBrevoContacts` query

Yes, that's exactly what I meant

File: .changeset/dirty-balloons-prove.md (hunk 8 above)
Comment (RainbowBunchie) [reply .changeset/dirty-balloons-prove.md]: [19ddf00](https://github.com/vivid-planet/comet-brevo-module/pull/52/commits/19ddf00d7ed8457a250ae66a85a49849e0de536f)

File: packages/admin/src/targetGroups/addContacts/AddContactsGridSelect.gql.ts (hunk 1 above)
Comment (RainbowBunchie) [reply packages/admin/src/targetGroups/addContacts/AddContactsGridSelect.gql.ts]: Renamed `brevoContactsInTargetGroup` to `manuallyAssignedBrevoContacts` and removed the `onlyShowManuallyAssignedContacts` argument.
Readded targetGroupId to the "normal" `brevoContacts` query

[4aaf79c](https://github.com/vivid-planet/comet-brevo-module/pull/52/commits/4aaf79c26659507fcd88f085e6e07f08fe7b34e8)

File: packages/admin/src/targetGroups/addContacts/AddContactsGridSelect.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (RainbowBunchie) [reply packages/admin/src/targetGroups/addContacts/AddContactsGridSelect.tsx]: removed [4aaf79c](https://github.com/vivid-planet/comet-brevo-module/pull/52/commits/4aaf79c26659507fcd88f085e6e07f08fe7b34e8)

File: packages/admin/src/targetGroups/TargetGroupsGrid.tsx [hunk 9]
```diff
@@ -128,9 +128,9 @@ export function TargetGroupsGrid({
     const dataGridProps = { ...useDataGridRemote(), ...usePersistentColumnState("TargetGroupsGrid") };
 
     const targetGroupContactsQuery = gql`
-        query TargetGroupContacts($targetGroupId: ID, $offset: Int, $limit: Int, $scope: EmailCampaignContentScopeInput!) {
-            brevoContacts(targetGroupId: $targetGroupId, offset: $offset, limit: $limit, scope: $scope) {
-                nodes {
+        query TargetGroupContacts($targetGroupId: ID!, $offset: Int, $limit: Int, $scope: EmailCampaignContentScopeInput!) {
+            brevoContacts(targetGroupId: $targetGroupId, offset: $offset, limit: $limit, scope: $scope) {               
+                 nodes {
```
Comment (johnnyomair) [inline packages/admin/src/targetGroups/TargetGroupsGrid.tsx:133]: Out of scope: Shouldn't this be formatted and linted so that these unnecessary whitespace changes don't happen?

**Review by johnnyomair — APPROVED** [review]

**Review by raphaelblum — APPROVED** [review]

File: packages/admin/src/targetGroups/TargetGroupsGrid.tsx (hunk 9 above)
Comment (raphaelblum) [reply packages/admin/src/targetGroups/TargetGroupsGrid.tsx:133]: it formats the whitespaces automatically when saving, and in the pipeline it fails when having wrong whitespaces: https://github.com/vivid-planet/comet-brevo-module/pull/54

The only thing I have just seen is that the husky precommit doesn't work correctly. I'll look into it.

**Review by thomasdax98 — APPROVED** [review]
